src/bgp.rs

```diff
@@ -0,0 +1,1049 @@
+    pub bgp_id: [u8; 4],
+    /// Optional Parameters Length: Total length of Optional Parameters.
+    pub opt_parm_len: u8,
+}
+
+impl OpenMsgLayout {
+    /// Length of the fixed part of an OPEN message in bytes.
+    pub const LEN: usize = 10;
+    const VERSION_OFFSET: usize = 0;
+    const MY_AS_OFFSET: usize = 1;
+    const HOLD_TIME_OFFSET: usize = 3;
+    const BGP_ID_OFFSET: usize = 5;
+    const OPT_PARM_LEN_OFFSET: usize = 9;
+}
+
+/// Initially fixed part of a BGP UPDATE message (RFC 4271, Section 4.3).
+#[repr(C, packed)]
+#[derive(Debug, Copy, Clone, Default)]
+#[cfg_attr(feature = "serde", derive(::serde::Serialize, ::serde::Deserialize))]
+pub struct UpdateInitialMsgLayout {
+    /// Withdrawn Routes Length: Total length of Withdrawn Routes.
+    pub withdrawn_routes_len: [u8; 2],
+}
+
+impl UpdateInitialMsgLayout {
+    /// Length of the initially fixed part of an UPDATE message in bytes.
+    pub const LEN: usize = 2;
+    const WITHDRAWN_ROUTES_LEN_OFFSET: usize = 0;
+}
+
+/// Fixed part of a BGP NOTIFICATION message (RFC 4271, Section 4.5).
+#[repr(C, packed)]
+#[derive(Debug, Copy, Clone, Default)]
+#[cfg_attr(feature = "serde", derive(::serde::Serialize, ::serde::Deserialize))]
+pub struct NotificationMsgLayout {
+    /// Error Code: Type of error.
+    pub error_code: u8,
+    /// Error Subcode: Specific information about the error.
+    pub error_subcode: u8,
+}
+
+impl NotificationMsgLayout {
+    /// Length of the fixed part of a NOTIFICATION message in bytes.
+    pub const LEN: usize = 2;
+    const ERROR_CODE_OFFSET: usize = 0;
+    const ERROR_SUBCODE_OFFSET: usize = 1;
+}
+
+/// Fixed part of a BGP ROUTE-REFRESH message (RFC 2918, Section 3).
+#[repr(C, packed)]
+#[derive(Debug, Copy, Clone, Default)]
+#[cfg_attr(feature = "serde", derive(::serde::Serialize, ::serde::Deserialize))]
+pub struct RouteRefreshMsgLayout {
+    /// Address Family Identifier (AFI)
+    pub afi: [u8; 2],
+    /// Reserved: Set to 0.
+    pub _reserved: u8,
+    /// Subsequent Address Family Identifier (SAFI).
+    pub safi: u8,
+}
+
+impl RouteRefreshMsgLayout {
+    /// Length of the fixed part of a ROUTE-REFRESH message in bytes.
+    pub const LEN: usize = 4;
+    const AFI_OFFSET: usize = 0;
+    const RES_OFFSET: usize = 2;
+    const SAFI_OFFSET: usize = 3;
+}
+
+/// BGP header and initially fixed part of its payload.
+#[repr(C, packed)]
+#[derive(Copy, Clone)]
+#[cfg_attr(feature = "serde", derive(::serde::Serialize, ::serde::Deserialize))]
+pub struct BgpHdr {
+    /// Marker: MUST be all ones (RFC 4271).
+    pub marker: [u8; 16],
+    /// Length: Total length of the BGP message in octets,
+    /// including the header. Stored in network byte order.
+    pub length: [u8; 2],
+    /// Type: Type code of the BGP message.
+    pub msg_type: u8,
+    /// Bytes for the message-specific fixed payload.
+    /// Sized by the largest possible fixed payload (`OpenMsgLayout::LEN`).
+    pub specific_payload_bytes: [u8; OpenMsgLayout::LEN],
```

Comment:
I would recommend changing the message data field to a union:

```suggestion
    pub data: BpgMsgUn,
```
(Check out the open [ICMP PR](https://github.com/vadorovsky/network-types/pull/48/files) for inspiration.)

Then, define a union within this module:
```
pub union BpgMsgUn {
    . . . <Various message structs>
}
```
(ref: [ICMP Union](https://github.com/vadorovsky/network-types/pull/48/files#diff-737a65f6f719c3326171fb235f40b29d7548b57d92fd3d4f54b0e5e175b8b482R443))